service/hub/server/server.go

```diff
@@ -0,0 +1,373 @@
+	ares := []AuthzResult{}
+	for _, scope := range ar.Scopes {
+		ai := &api.AuthRequestInfo{
+			Account:    ar.Account,
+			Type:       scope.Type,
+			Name:       scope.Name,
+			Service:    ar.Service,
+			IP:         ar.RemoteIP,
+			Actions:    scope.Actions,
+			Labels:     ar.Labels,
+			Kubeconfig: string(ar.Password),
+		}
+		actions, err := as.authorizeScope(ai)
+		if err != nil {
+			return nil, err
+		}
+		ares = append(ares, AuthzResult{scope: scope, autorizedActions: actions})
+	}
+	return ares, nil
+}
+
+// CreateToken https://github.com/docker/distribution/blob/master/docs/spec/auth/token.md#example
+func (as *AuthServer) CreateToken(ar *AuthRequest, ares []AuthzResult) (string, error) {
+	now := time.Now().Unix()
+	tc := &as.config.Token
+
+	// Sign something dummy to find out which algorithm is used.
+	_, sigAlg, err := tc.privateKey.Sign(strings.NewReader("dummy"), 0)
+	if err != nil {
+		return "", fmt.Errorf("failed to sign: %s", err)
+	}
+	header := token.Header{
+		Type:       "JWT",
+		SigningAlg: sigAlg,
+		KeyID:      tc.publicKey.KeyID(),
+	}
+	headerJSON, err := json.Marshal(header)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal header: %s", err)
+	}
+
+	claims := token.ClaimSet{
+		Issuer:     tc.Issuer,
+		Subject:    ar.Account,
+		Audience:   ar.Service,
+		NotBefore:  now - 10,
+		IssuedAt:   now,
+		Expiration: now + tc.Expiration,
+		JWTID:      fmt.Sprintf("%d", rand.Int63()),
+		Access:     []*token.ResourceActions{},
+	}
+	for _, a := range ares {
+		ra := &token.ResourceActions{
+			Type:    a.scope.Type,
+			Name:    a.scope.Name,
+			Actions: a.autorizedActions,
+		}
+		if ra.Actions == nil {
+			ra.Actions = []string{}
+		}
+		sort.Strings(ra.Actions)
+		claims.Access = append(claims.Access, ra)
+	}
+	claimsJSON, err := json.Marshal(claims)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal claims: %s", err)
+	}
+
+	payload := fmt.Sprintf("%s%s%s", joseBase64UrlEncode(headerJSON), token.TokenSeparator, joseBase64UrlEncode(claimsJSON))
+
+	sig, sigAlg2, err := tc.privateKey.Sign(strings.NewReader(payload), 0)
+	if err != nil || sigAlg2 != sigAlg {
+		return "", fmt.Errorf("failed to sign token: %s", err)
+	}
+	glog.Infof("New token for %s %+v: %s", *ar, ar.Labels, claimsJSON)
+	return fmt.Sprintf("%s%s%s", payload, token.TokenSeparator, joseBase64UrlEncode(sig)), nil
+}
+
+func (as *AuthServer) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	glog.V(3).Infof("Request: %+v", req)
+	pathPrefix := as.config.Server.PathPrefix
+	if as.config.Server.HSTS {
+		rw.Header().Add("Strict-Transport-Security", "max-age=63072000; includeSubDomains")
+	}
+	switch {
+	case req.URL.Path == pathPrefix+"/":
+		as.doIndex(rw, req)
+	case req.URL.Path == pathPrefix+"/auth":
+		as.doAuth(rw, req)
+	default:
+		http.Error(rw, "Not found", http.StatusNotFound)
+		return
+	}
+}
+
+func (as *AuthServer) doIndex(rw http.ResponseWriter, req *http.Request) {
```

Comment:
A page to show token?

Reply:
return auth server's issuer, it should be add at docker v2 registry config.auth.token.issuer